### PR TITLE
resource_manager/client: introduce watch resource group

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #6509

### What is changed and how does it work?
In the past, we didn't need to update the configuration of the group in a timely manner because when we only had the token bucket function, we only needed to get tokens from the server. But after joining the runaway query, we need to get a new configuration in time.

```commit-message
```

### Check List

Tests

- Integration test

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

**Comment by CabinfeverB** [comment]
cc @HuSharp @JmPotato

File: client/resource_group/controller/controller.go [hunk 1]
```diff
@@ -58,13 +62,14 @@ type ResourceGroupKVInterceptor interface {
...
 type ResourceGroupProvider interface {
-	ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error)
 	GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error)
 	AddResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
 	ModifyResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
 	DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error)
 	AcquireTokenBuckets(ctx context.Context, request *rmpb.TokenBucketsRequest) ([]*rmpb.TokenBucketResponse, error)
 	LoadGlobalConfig(ctx context.Context, names []string, configPath string) ([]pd.GlobalConfigItem, int64, error)
+	LoadResourcrGroups(ctx context.Context) ([]*rmpb.ResourceGroup, int64, error)
+	Watch(ctx context.Context, key []byte, opts ...pd.OpOption) (chan []*meta_storagepb.Event, error)
```
Comment (HuSharp) [inline client/resource_group/controller/controller.go:72]: Do we need to keep consistent with ResourceGroupClient?

File: client/resource_manager_client.go [hunk 2]
```diff
@@ -158,12 +162,31 @@ func (c *client) DeleteResourceGroup(ctx context.Context, resourceGroupName stri
...
+		group := &rmpb.ResourceGroup{}
+		if err := proto.Unmarshal(item.Value, group); err != nil {
+			continue
+		}
+		groups = append(groups, group)
+	}
+	return groups, resp.Header.Revision, nil
+}
+
 // WatchResourceGroup [just for TEST] watches resource groups changes.
```
Comment (HuSharp) [inline client/resource_manager_client.go:184]: Can remove `[just for TEST]`

Comment (CabinfeverB) [reply client/resource_manager_client.go:184]: This function is still used for testing purposes

File: client/resource_group/controller/controller.go (hunk 1 above)
Comment (CabinfeverB) [reply client/resource_group/controller/controller.go:72]: Can we can add the `Watch` into ResourceGroupClient interface?

File: client/resource_group/controller/controller.go [hunk 3]
```diff
@@ -58,13 +62,14 @@ type ResourceGroupKVInterceptor interface {
...
 // ResourceGroupProvider provides some api to interact with resource manager server。
 type ResourceGroupProvider interface {
-	ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error)
 	GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error)
 	AddResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
 	ModifyResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
 	DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error)
 	AcquireTokenBuckets(ctx context.Context, request *rmpb.TokenBucketsRequest) ([]*rmpb.TokenBucketResponse, error)
 	LoadGlobalConfig(ctx context.Context, names []string, configPath string) ([]pd.GlobalConfigItem, int64, error)
+	LoadResourcrGroups(ctx context.Context) ([]*rmpb.ResourceGroup, int64, error)
```
Comment (Connor1996) [inline client/resource_group/controller/controller.go:71]: Resourcr -> Resource

**Review by Connor1996 — APPROVED** [review]
LGTM

File: client/resource_group/controller/controller.go [hunk 4]
```diff
@@ -289,23 +338,9 @@ func (c *ResourceGroupsController) tryGetResourceGroup(ctx context.Context, name
...
-func (c *ResourceGroupsController) cleanUpResourceGroup(ctx context.Context) error {
-	groups, err := c.provider.ListResourceGroups(ctx)
-	if err != nil {
-		return errs.ErrClientListResourceGroup.FastGenByArgs(err.Error())
-	}
-	latestGroups := make(map[string]struct{})
-	for _, group := range groups {
-		latestGroups[group.GetName()] = struct{}{}
-	}
+func (c *ResourceGroupsController) cleanUpResourceGroup() {
```
Comment (Connor1996) [inline client/resource_group/controller/controller.go:341]: help add a `GetResourceGrouup()` by the way

File: client/resource_group/controller/controller.go [hunk 5]
```diff
@@ -732,7 +773,7 @@ func (gc *groupCostController) updateAvgRUPerSec() {
 		if !gc.calcAvg(counter, getRUValueFromConsumption(gc.run.consumption, typ)) {
 			continue
 		}
-		log.Debug("[resource group controller] update avg ru per sec", zap.String("name", gc.Name), zap.String("type", rmpb.RequestUnitType_name[int32(typ)]), zap.Float64("avgRUPerSec", counter.avgRUPerSec))
+		log.Debug("[resource group controller] update avg ru per sec", zap.String("name", gc.name), zap.String("type", rmpb.RequestUnitType_name[int32(typ)]), zap.Float64("avgRUPerSec", counter.avgRUPerSec))
```
Comment (HuSharp) [inline client/resource_group/controller/controller.go:776]: ditto

File: client/resource_group/controller/controller.go [hunk 6]
```diff
@@ -718,7 +759,7 @@ func (gc *groupCostController) updateAvgRaWResourcePerSec() {
 		if !gc.calcAvg(counter, getRawResourceValueFromConsumption(gc.run.consumption, typ)) {
 			continue
 		}
-		log.Debug("[resource group controller] update avg raw resource per sec", zap.String("name", gc.Name), zap.String("type", rmpb.RawResourceType_name[int32(typ)]), zap.Float64("avgRUPerSec", counter.avgRUPerSec))
+		log.Debug("[resource group controller] update avg raw resource per sec", zap.String("name", gc.name), zap.String("type", rmpb.RawResourceType_name[int32(typ)]), zap.Float64("avgRUPerSec", counter.avgRUPerSec))
```
Comment (HuSharp) [inline client/resource_group/controller/controller.go:762]:
```suggestion
		log.Debug("[resource group controller] update avg raw resource per sec", zap.String("name", gc.name), zap.String("type", rmpb.RawResourceType_name[int32(typ)]), zap.Float64("avg-ru-per-sec", counter.avgRUPerSec))
```
Do we need to keep consistent with this https://github.com/tikv/pd/pull/6450/files#diff-ed72d626078622ff5c8e9b3f7d1876c4b3d6c648a4bc833785ca312da718aa07R319

**Review by HuSharp — APPROVED** [review]
rest LGTM!

**Pull request merged**